### ISSUE TITLE
Consider pre-packaging the "PreMaterialize" pattern (we could late optimize it too)

### ISSUE THREAD

**Issue by ktoso** [akka/akka#23895]
Pretty sure not many people can ad-hoc whip up this pattern to get a materialized value powered source to emit an entity yet it's very useful in practice:

example:

```scala
    val introspector = path("introspector") {
      concat(
        path(Segment / "inspect") { id ⇒
          val refSource = Source.actorRef[ReplicatedIntrospector.Event](Int.MaxValue, OverflowStrategy.fail)
            .map(_.render)

          // TODO "pre-materialize pattern"
          val (ref, pub) =refSource.toMat(Sink.asPublisher(true))(Keep.both).run()
          val s = Source.fromPublisher(pub)
            .intersperse("\n")
            .map(ByteString(_))
          // end of "pre-materialize pattern"

          introspectorProxy.tell(ReplicatedIntrospector.Inspect, ref)
          
          complete(HttpResponse(entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, s)))
        }
      )
    }

```

I think we should be able to pre-build such a thing perhaps, document it as pre-materialize something and allow users to use it thanks to that rather than have them re-invent this pattern

**Comment by ktoso** [comment]
Example is HTTP, though the use case is "materialized value powered source" so opened for streams

**Comment by ktoso** [comment]
Let's do it!

**Comment by svezfaz** [comment]
@ktoso could we start by packing this in a combinator? Could it even be called `run()`, to be more in line with the nomenclature in use?

Your code would become something along the lines of:

    val refSource = Source.actorRef[String](Int.MaxValue, OverflowStrategy.fail)

    val (s, ref) = refSource.run()

    introspectorProxy.tell(ReplicatedIntrospector.Inspect, ref)
          
    complete(HttpResponse(entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, s)))

**Comment by ktoso** [comment]
I think just a run() is too "nice looking", it's an expensive operation and very different than the usual case. I think we can start by putting the Publisher/Subscriber dance into an operator, eventually making it more optimal, however we need to figure out a name and way to do so without inviting people to do this "too often", as it's a heavy thing in general, also, could leak to resource leaks etc...

**Comment by svezfaz** [comment]
Point taken. I was looking for something more friendly than `preMaterialize`, but can't for now.
I'll get a draft out, if a better name comes up we can amend it.

**Comment by ktoso** [comment]
Yeah if we come up with a better name that’s easy enough to change in the
PR.
Thanks!

-- 
Cheers,
Konrad 'ktoso <http://kto.so>' Malawski
Akka <http://akka.io/> @ Lightbend <http://lightbend.com/>

On February 1, 2018 at 22:45:01, Stefano Bonetti (notifications@github.com)
wrote:

Point taken. I was looking for something more friendly than preMaterialize,
but can't for now.
I'll get a draft out, if a better name comes up we can amend it.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/akka/akka/issues/23895#issuecomment-362269816>, or mute
the thread

**Comment by johanandren** [comment]
I think this was incorrectly marked as in 2.5.12 while it was actually already in 2.5.11 ?

**Comment by ktoso** [comment]
Yes, thanks for the catch
